src/llmq/quorums_dkgsessionhandler.cpp

```diff
@@ -0,0 +1,577 @@
+// Copyright (c) 2018-2021 The Dash Core developers
+// Copyright (c) 2022 The PIVX developers
+// Distributed under the MIT/X11 software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include "llmq/quorums_dkgsessionhandler.h"
+
+#include "activemasternode.h"
+#include "chainparams.h"
+#include "llmq/quorums_blockprocessor.h"
+#include "llmq/quorums_init.h"
+#include "llmq/quorums_utils.h"
+#include "net_processing.h"
+#include "shutdown.h"
+#include "util/threadnames.h"
+#include "validation.h"
+
+namespace llmq
+{
+
+CDKGPendingMessages::CDKGPendingMessages(size_t _maxMessagesPerNode) :
+    maxMessagesPerNode(_maxMessagesPerNode)
+{
+}
+
+void CDKGPendingMessages::PushPendingMessage(NodeId from, CDataStream& vRecv, int invType)
+{
+    // this will also consume the data, even if we bail out early
+    auto pm = std::make_shared<CDataStream>(std::move(vRecv));
+
+    {
+        LOCK(cs);
+
+        if (messagesPerNode[from] >= maxMessagesPerNode) {
+            // TODO ban?
+            LogPrint(BCLog::NET, "CDKGPendingMessages::%s -- too many messages, peer=%d\n", __func__, from);
+            return;
+        }
+        messagesPerNode[from]++;
+    }
+
+    CHashWriter hw(SER_GETHASH, 0);
+    hw.write(pm->data(), pm->size());
+    uint256 hash = hw.GetHash();
+
+    LOCK2(cs_main, cs);
+
+    g_connman->RemoveAskFor(hash, invType);
+
+    if (!seenMessages.emplace(hash).second) {
+        LogPrint(BCLog::NET, "CDKGPendingMessages::%s -- already seen %s, peer=%d\n", __func__, hash.ToString(), from);
+        return;
+    }
+
+    pendingMessages.emplace_back(std::make_pair(from, std::move(pm)));
+}
+
+std::list<CDKGPendingMessages::BinaryMessage> CDKGPendingMessages::PopPendingMessages(size_t maxCount)
+{
+    LOCK(cs);
+
+    std::list<BinaryMessage> ret;
+    while (!pendingMessages.empty() && ret.size() < maxCount) {
+        ret.emplace_back(std::move(pendingMessages.front()));
+        pendingMessages.pop_front();
+    }
+
+    return std::move(ret);
+}
+
+bool CDKGPendingMessages::HasSeen(const uint256& hash) const
+{
+    LOCK(cs);
+    return seenMessages.count(hash) != 0;
+}
+
+void CDKGPendingMessages::Clear()
+{
+    LOCK(cs);
+    pendingMessages.clear();
+    messagesPerNode.clear();
+    seenMessages.clear();
+}
+
+//////
+
+CDKGSessionHandler::CDKGSessionHandler(const Consensus::LLMQParams& _params, CEvoDB& _evoDb, ctpl::thread_pool& _messageHandlerPool, CBLSWorker& _blsWorker, CDKGSessionManager& _dkgManager) :
+    params(_params),
+    evoDb(_evoDb),
+    messageHandlerPool(_messageHandlerPool),
```

Comment:
This `messageHandlerPool` seems to be set but never used

Reply:
True. Have reworked the thread handling in 7517ef097989e7279d862163f7781bf87f209d67